expression/new_constant_propagation.go

```diff
@@ -0,0 +1,162 @@
+	if s.constfalse {
+		return []Expression{&Constant{
+			Value:   types.NewDatum(false),
+			RetType: types.NewFieldType(mysql.TypeTiny),
+		}}
+	}
+
+	idx := 0
+	for i := 0; i < len(s.data); i++ {
+		if !s.tombstone[i] {
+			s.data[idx] = s.data[i]
+			idx++
+		}
+	}
+	return s.data[:idx]
+}
+
+func (s *exprSet) SetConstFalse() {
+	s.constfalse = true
+}
+
+func newExprSet(conditions []Expression) *exprSet {
+	var exprs exprSet
+	exprs.data = make([]Expression, 0, len(conditions))
+	exprs.tombstone = make([]bool, 0, len(conditions))
+	exprs.index = make(map[string]struct{}, len(conditions))
+	for _, v := range conditions {
+		exprs.Append(v)
+	}
+	return &exprs
+}
+
+type pgSolver2 struct{}
+
+// PropagateConstant propagate constant values of deterministic predicates in a condition.
+func (pgSolver2) PropagateConstant(ctx sessionctx.Context, conditions []Expression) []Expression {
+	exprs := newExprSet(conditions)
+	fixPoint(ctx, exprs)
+	return exprs.Slice()
+}
+
+// fixPoint is the core of the constant propagation algorithm.
+// It will iterate the expression set over and over again, pick two expressions,
+// apply one to another.
+// If new conditions can be infered, they will be append into the expression set.
+// Until no more conditions can be infered from the set, the algorithm finish.
+func fixPoint(ctx sessionctx.Context, exprs *exprSet) {
+	for {
+		saveLen := len(exprs.data)
+		iterOnce(ctx, exprs)
+		if saveLen == len(exprs.data) {
+			break
+		}
+	}
+	return
+}
+
+// iterOnce picks two expressions from the set, try to propagate new conditions from them.
+func iterOnce(ctx sessionctx.Context, exprs *exprSet) {
+	for i := 0; i < len(exprs.data); i++ {
+		if exprs.tombstone[i] {
+			continue
+		}
+		for j := 0; j < len(exprs.data); j++ {
+			if exprs.tombstone[j] {
+				continue
+			}
+			if i == j {
+				continue
+			}
+			solve(ctx, i, j, exprs)
+		}
+	}
+}
+
+// solve uses exprs[i] exprs[j] to propagate new conditions.
+func solve(ctx sessionctx.Context, i, j int, exprs *exprSet) {
+	for _, rule := range rules {
+		rule(ctx.GetSessionVars().StmtCtx, i, j, exprs)
+	}
+}
+
+type constantPropagateRule func(ctx *stmtctx.StatementContext, i, j int, exprs *exprSet)
+
+var rules = []constantPropagateRule{
+	ruleConstantFalse,
+	ruleColumnEQConst,
+}
+
+// ruleConstantFalse propagates from CNF condition that false plus anything returns false.
+// false, a = 1, b = c ... => false
+func ruleConstantFalse(ctx *stmtctx.StatementContext, i, j int, exprs *exprSet) {
+	cond := exprs.data[i]
+	if cons, ok := cond.(*Constant); ok {
+		switch cons.RetType.Tp {
+		case mysql.TypeTiny, mysql.TypeShort, mysql.TypeLong, mysql.TypeLonglong:
+			if cons.Value.GetInt64() == 0 {
+				exprs.SetConstFalse()
+			}
+		}
+	}
+}
+
+// ruleColumnEQConst propagates the "column = const" condition.
+// "a = 3, b = a, c = a, d = b" => "a = 3, b = 3, c = 3, d = 3"
+func ruleColumnEQConst(ctx *stmtctx.StatementContext, i, j int, exprs *exprSet) {
```

Comment:
Seems `ColumnEQ` stands for `col1 = col2` while `ConstEQ` stands for `col = const` in current code base.

Reply:
You mean in the old constant_propagation.go ? I don't care the naming details.
The algorithm should works as a flexible framework, new rules can be add or remove.
For example, if the old code already infer `col1 = col2`, `col = const`, then those rules can be skipped.

We can define a solver like that:

```
type propConstSolver []constantPropagateRule

func newPropConstSolver(rules []constantPropagateRule) PropagateConstantSolver {
    return rules
}

// if I want to infer eq condition
newPropConstSolver([]{ruleColumnEQConst, ruleConstantFalse}) 

// I want to infer gt condition
newPropConstSolver([]{ruleConstGTColumn}) 

// The real thing I need, to handle partition prunning like `to_days` function
newPropConstSolver([]{ruleMonotonicFunction}) 
```

New rules will be add later.